File: frontend/src/types/stats.ts
/**
 * 统计相关类型定义
 *
 * 契约真相源是后端 schemas/stats.py（→ api.generated.ts）；
 * 底部 SameShape 锚点锁住手写类型与生成物逐字段一致。
 */

import type { components } from '@/types/api.generated'
import type { RunStatus } from './run'

/**
 * 运行核心指标
 */
export interface RunMetrics {
  total_runs: number
  success_count: number
  failed_count: number
  hitl_count: number
  avg_duration_ms: number
  success_rate: number
}

/**
 * 每日趋势数据
 */
export interface DailyTrend {
  date: string
  total_count: number
  success_count: number
  failed_count: number
}

/**
 * 运行列表项
 */
export interface RunListItem {
  run_id: string
  thread_id: string
  user_id: string | null
  user_name: string | null
  /** 可空 str（跟随契约）：NULL=路由决策前终止；后端对存量 "router" 值宽容 */
  mode: string | null
  status: RunStatus
  duration_ms: number | null
  created_at: string
  completed_at: string | null
}

/**
 * 运行统计响应
 */
export interface RunStatsResponse {
  is_admin: boolean

  // 今日 token 用量与每用户日配额（null = 不限量）
  today_tokens: number
  daily_token_quota: number | null
  metrics: RunMetrics
  trends: DailyTrend[]
  runs: RunListItem[]
  total_runs_count: number
  limit: number
  offset: number
}

// ============================================
// REST 契约锚点（范式沿 types/events.ts 的 SameShape / services/stats.ts）
// ============================================

/** 双向相等：手写类型与后端生成类型**逐字段一致**（含可选性与 null）。 */
type SameShape<A, B> = [A] extends [B] ? ([B] extends [A] ? true : false) : false
type Assert<T extends true> = T

type Schemas = components['schemas']
type _RunMetrics = Assert<SameShape<RunMetrics, Schemas['RunMetrics']>>
type _DailyTrend = Assert<SameShape<DailyTrend, Schemas['DailyTrend']>>
type _RunListItem = Assert<SameShape<RunListItem, Schemas['RunListItem']>>
type _RunStatsResponse = Assert<SameShape<RunStatsResponse, Schemas['RunStatsResponse']>>

/** 上面这组断言只做编译期校验，导出以免被 noUnusedLocals 误报 */
export type StatsConformanceAnchors = [
  _RunMetrics,
  _DailyTrend,
  _RunListItem,
  _RunStatsResponse,
]

File: backend/schemas/stats.py
"""
统计 API 响应模型

用于运行统计和趋势数据返回。
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from models.enums import RunStatus


class RunMetrics(BaseModel):
    """运行核心指标"""

    total_runs: int  # 总运行数
    success_count: int  # 成功数
    failed_count: int  # 失败数
    hitl_count: int  # HITL 拦截次数
    avg_duration_ms: float  # 平均耗时（毫秒）
    success_rate: float  # 成功率（百分比）


class DailyTrend(BaseModel):
    """每日趋势数据"""

    date: str  # 日期 YYYY-MM-DD
    total_count: int  # 当天总运行数
    success_count: int  # 当天成功数
    failed_count: int  # 当天失败数


class RunListItem(BaseModel):
    """运行列表项

    按「恒序列化、值可空的字段不写默认值」约定，user_id/user_name 无默认值——
    crud 组装的 dict 恒带全部键，默认值只会让 OpenAPI 契约退化成 optional。
    """

    run_id: str
    thread_id: str
    user_id: str | None  # 仅 admin 可见
    user_name: str | None  # 仅 admin 可见
    mode: Literal["simple", "complex"]
    status: RunStatus
    duration_ms: int | None
    created_at: datetime
    completed_at: datetime | None


class RunStatsResponse(BaseModel):
    """运行统计响应"""

    # 权限标识
    is_admin: bool

    # 今日 token 用量与每用户日配额（quota 为 null 表示不限量）
    today_tokens: int
    daily_token_quota: int | None

    # 核心指标
    metrics: RunMetrics

    # 趋势数据（按天统计，最近 7 天）
    trends: list[DailyTrend]

    # 运行列表（分页）
    runs: list[RunListItem]

    # 分页信息
    total_runs_count: int  # 总记录数（用于分页）
    limit: int
    offset: int


class TokensTodayResponse(BaseModel):
    """今日 token 用量 + 全局配额（底栏轻量轮询专供）。

    约定：响应模型里「恒序列化、值可空」的字段不写默认值——写了默认值会被
    OpenAPI 降级成 optional，而 FastAPI 运行时其实恒有键（null 也是键）；
    无默认值才能生成 `required + nullable` 的准确契约。
    """

    today_tokens: int
    daily_token_quota: int | None
